document-store/src/main/java/org/hypertrace/core/documentstore/mongo/query/parser/MongoLetClauseBuilder.java

```diff
@@ -0,0 +1,104 @@
+import org.hypertrace.core.documentstore.expression.impl.RelationalExpression;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.parser.FilterTypeExpressionVisitor;
+import org.hypertrace.core.documentstore.parser.SelectTypeExpressionVisitor;
+
+// Visitor for building the let clause in lookup stage from filter expressions
+class MongoLetClauseBuilder implements FilterTypeExpressionVisitor {
+
+  private final String subQueryAlias;
+
+  MongoLetClauseBuilder(String subQueryAlias) {
+    this.subQueryAlias = subQueryAlias;
+  }
+
+  @Override
+  public BasicDBObject visit(LogicalExpression expression) {
+    BasicDBObject letClause = new BasicDBObject();
+    for (FilterTypeExpression operand : expression.getOperands()) {
+      letClause.putAll((Map<String, Object>) operand.accept(this));
+    }
+    return letClause;
+  }
+
+  @Override
+  public BasicDBObject visit(RelationalExpression expression) {
+    BasicDBObject letClause = new BasicDBObject();
+    letClause.putAll(
+        (Map<String, Object>)
+            expression.getLhs().accept(new MongoLetClauseSelectTypeExpressionVisitor()));
+    letClause.putAll(
+        (Map<String, Object>)
+            expression.getRhs().accept(new MongoLetClauseSelectTypeExpressionVisitor()));
+    return letClause;
+  }
+
+  private class MongoLetClauseSelectTypeExpressionVisitor implements SelectTypeExpressionVisitor {
+    @Override
+    public BasicDBObject visit(AggregateExpression expression) {
+      return new BasicDBObject();
+    }
+
+    @Override
+    public BasicDBObject visit(ConstantExpression expression) {
+      return new BasicDBObject();
+    }
+
+    @Override
+    public BasicDBObject visit(ConstantExpression.DocumentConstantExpression expression) {
+      return new BasicDBObject();
+    }
+
+    @Override
+    public BasicDBObject visit(FunctionExpression expression) {
+      return new BasicDBObject();
+    }
+
+    @Override
+    public BasicDBObject visit(IdentifierExpression expression) {
+      return new BasicDBObject();
+    }
+
+    @Override
+    public BasicDBObject visit(AliasedIdentifierExpression expression) {
+      return createLetClause(expression, subQueryAlias);
+    }
+  }
+
+  @Override
+  public BasicDBObject visit(KeyExpression expression) {
+    return new BasicDBObject();
+  }
+
+  @Override
+  public BasicDBObject visit(ArrayRelationalFilterExpression expression) {
+    return new BasicDBObject();
+  }
+
+  @Override
+  public BasicDBObject visit(DocumentArrayFilterExpression expression) {
+    return new BasicDBObject();
+  }
+
+  private BasicDBObject createLetClause(
+      AliasedIdentifierExpression aliasedExpression, String subQueryAlias) {
+    BasicDBObject letClause = new BasicDBObject();
+    if (aliasedExpression.getContextAlias().equals(subQueryAlias)) {
+      letClause.put(aliasedExpression.getName(), "$" + aliasedExpression.getName());
```

Comment:
What happens if the `AliasedExpression` contains a nested field `like "a.b"`. 
Does it need escaping/encoding? 

It would be beneficial to add an integration for the same.

Reply:
The field name of the aliasedExpression is given by the user and if we can make an assumption that the user won't give a dot in the field name then we should be fine as we are only using the field name here and not the context alias.

Reply:
I'm a bit confused. Does it allow me to pick the value of "a.b" and alias it as "c"? 

In other words, are you saying that the alias can't contain a dot, but the identifier name can?

Reply:
Sorry, my bad, I misunderstood your previous comment. I went through it again and it made sense this time. I've updated the implementation to encode the variable name so that it can handle nested fields also. I've also added an integration test for it.